lib/lightning/credentials/oauth_token.ex

```diff
@@ -0,0 +1,163 @@
+defmodule Lightning.Credentials.OauthToken do
+  @moduledoc """
+  Schema and functions for managing OAuth tokens. This module handles the storage and
+  validation of OAuth tokens, allowing multiple credentials to share the same token
+  when they have identical scope sets.
+  """
+  use Lightning.Schema
+  import Ecto.Query
+
+  alias Lightning.Accounts.User
+  alias Lightning.Credentials
+  alias Lightning.Credentials.Credential
+  alias Lightning.Credentials.OauthClient
+
+  @type t :: %__MODULE__{
+          id: Ecto.UUID.t() | nil,
+          body: map(),
+          scopes: [String.t()],
+          oauth_client_id: Ecto.UUID.t() | nil,
+          oauth_client: OauthClient.t() | nil,
+          user_id: Ecto.UUID.t() | nil,
+          user: User.t() | nil,
+          credentials: [Credential.t()] | nil,
+          inserted_at: DateTime.t() | nil,
+          updated_at: DateTime.t() | nil
+        }
+
+  schema "oauth_tokens" do
+    field :body, Lightning.Encrypted.Map, redact: true
+    field :scopes, {:array, :string}
+
+    belongs_to :oauth_client, OauthClient
+    belongs_to :user, User
+    has_many :credentials, Credential
+
+    timestamps()
+  end
+
+  @doc """
+  Creates a changeset for an OAuth token.
+  """
+  def changeset(oauth_token, attrs) do
+    oauth_token
+    |> cast(attrs, [:body, :scopes, :oauth_client_id, :user_id])
+    |> validate_required([:body, :scopes, :oauth_client_id, :user_id])
+    |> assoc_constraint(:oauth_client)
+    |> assoc_constraint(:user)
+    |> validate_oauth_body()
+  end
+
+  @doc """
+  Creates a changeset for updating token data.
+  """
+  def update_token_changeset(oauth_token, new_token) do
+    scopes =
+      case extract_scopes(new_token) do
+        {:ok, scopes} -> scopes
+        :error -> nil
+      end
+
+    cast(oauth_token, %{body: new_token, scopes: scopes}, [:body, :scopes])
+    |> validate_required([:body, :scopes])
+    |> validate_oauth_body()
+  end
+
+  @doc """
+  Finds or creates an OAuth token for the given user, client and scope set.
+  Returns {:ok, token} if successful, {:error, changeset} if validation fails.
+  """
+  def find_or_create_for_scopes(user_id, oauth_client_id, scopes, tokens)
+      when is_list(scopes) do
+    case find_by_scopes(user_id, oauth_client_id, scopes) do
+      nil ->
+        %__MODULE__{}
+        |> changeset(%{
+          user_id: user_id,
+          oauth_client_id: oauth_client_id,
+          scopes: scopes,
+          body: tokens
+        })
+        |> Lightning.Repo.insert()
```

Comment:
You probably wanted to move it to a context module. For this and the operations bellow to avoid the insert and querying in the Schema module.

Reply:
This is done, nice catch, thanks